File: Analyzer/cpp/lib/Manager/DBMgr/LinkMgr.cpp
#include "Manager/DBMgr/LinkMgr.h"

#include "Utils/DBOperations.h"

using namespace std;

/*
 * Implementation of LinkMgr
 */

LinkMgr::LinkMgr(PathMgr *path_mgr, DBMgr *db_mgr) : TBLMgr(db_mgr) {
	this->path_mgr = path_mgr;
	this->db_mgr = db_mgr;
	string tbl_create_link = path_mgr->getTblCreateLink();
	string tbl_create_link_sql = loadSql(tbl_create_link);
	db_mgr->execSimple(tbl_create_link_sql);
}

LinkMgr::~LinkMgr() {
	// Need to do nothing at present.
}

bool LinkMgr::insert(LinkRow &row) {
	SqlParams sql_params = {row.target_file, row.link_list, row.ir_list}; 
	string tbl_insert_link = path_mgr->getTblInsertLink();
	string tbl_insert_link_sql = loadSql(tbl_insert_link);
	return db_mgr->execSql(tbl_insert_link_sql, sql_params);
}

bool LinkMgr::insertBatch(vector<LinkRow> &row_vec) {
	vector<SqlParams> params_list;
	params_list.reserve(row_vec.size());

	for (size_t row_idx = 0; row_idx < row_vec.size(); row_idx++) {
		LinkRow &row = row_vec[row_idx];
		params_list.push_back({row.target_file, row.link_list, row.ir_list});
	}
	string tbl_insert_link = path_mgr->getTblInsertLink();
	string tbl_insert_link_sql = loadSql(tbl_insert_link);
	return db_mgr->execSqlBatch(tbl_insert_link_sql, params_list);
}

vector<LinkRow> LinkMgr::selectAll() {
	string tbl_select_link = path_mgr->getTblSelectLink();
	string tbl_select_link_sql = loadSql(tbl_select_link);
	vector<LinkRow> results;

	sqlite3_stmt * stmt = db_mgr->prepareQuery(tbl_select_link_sql, {});
	if (!stmt) {
		return results;
	}
	while (sqlite3_step(stmt) == SQLITE_ROW) {
		LinkRow row;
		row.target_file = get_column_text(stmt, 0);
		row.link_list = get_column_text(stmt, 1);
		row.ir_list = get_column_text(stmt, 2);

		results.push_back(row);
	}

	db_mgr->finalizeQuery(stmt);
	return results;
}

File: Analyzer/python/RaceMinerLib/Manager/DBMgr/LinkMgr.py
import logging
from dataclasses import dataclass
from typing import List

from RaceMinerLib.Manager.DBMgr import DBMgr
from RaceMinerLib.Manager.DBMgr import TBLMgr
from RaceMinerLib.Manager.PathMgr import PathMgr


@dataclass
class LinkRow:
    target_file: str
    link_list: str
    ir_list: str


class LinkMgr(TBLMgr):
    def __init__(self, path_mgr, db_mgr):
        super().__init__(db_mgr)

        self.path_mgr = path_mgr
        self.db_mgr = db_mgr

        self.logger = logging.getLogger(self.__class__.__name__)
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                "[%(asctime)s][%(levelname)s] %(message)s"
            )
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)
            self.logger.setLevel(logging.INFO)

        sql_path = self.path_mgr.get_tbl_create_link()
        sql = self.load_sql(sql_path)

        if sql:
            self.db_mgr.exec_simple(sql)
            self.logger.info("Link table initialized")
        else:
            self.logger.error("Failed to initialize Link table")

    def insert(self, row: LinkRow) -> bool:
        sql_path = self.path_mgr.get_tbl_insert_link()
        sql = self.load_sql(sql_path)

        if not sql:
            return False

        params = [
            row.target_file, row.link_list, row.ir_list
        ]

        return self.db_mgr.exec_sql(sql, params)

    def insert_batch(self, row_list: List[LinkRow]) -> bool:
        if not row_list:
            return True

        sql_path = self.path_mgr.get_tbl_insert_link()
        sql = self.load_sql(sql_path)

        if not sql:
            return False

        params_list = [
            [
                r.target_file, r.link_list, r.ir_list
            ]
            for r in row_list
        ]

        return self.db_mgr.exec_sql_batch(sql, params_list)

    def select_all(self) -> List[LinkRow]:
        sql_path = self.path_mgr.get_tbl_select_link()
        sql = self.load_sql(sql_path)

        results: List[LinkRow] = []

        if not sql:
            return results

        cur = self.db_mgr.prepare_query(sql, [])
        if not cur:
            return results

        try:
            rows = cur.fetchall()

            for r in rows:
                row = LinkRow(
                    target_file=r[1],
                    link_list=r[2],
                    ir_list=r[3],
                )
                results.append(row)

        except Exception as e:
            self.logger.error(f"[select_all] {e}")

        finally:
            self.db_mgr.finalize_query(cur)

        return results
